rllib/offline/estimators/qreg_torch_model.py

```diff
@@ -0,0 +1,219 @@
+        self.n_iters = n_iters
+        self.lr = lr
+        self.delta = delta
+        self.clip_grad_norm = clip_grad_norm
+        self.batch_size = batch_size
+        self.tau = tau
+        self.optimizer = torch.optim.Adam(self.q_model.variables(), self.lr)
+        initializer = get_initializer("xavier_uniform", framework="torch")
+
+        def f(m):
+            if isinstance(m, nn.Linear):
+                initializer(m.weight)
+
+        self.initializer = f
+
+    def reset(self) -> None:
+        """Resets/Reinintializes the model weights."""
+        self.q_model.apply(self.initializer)
+
+    def train_q(self, batch: SampleBatch) -> TensorType:
+        """Trains self.q_model using Q-Reg loss on given batch.
+
+        Args:
+            batch: A SampleBatch of episodes to train on
+
+        Returns:
+            A list of losses for each training iteration
+        """
+        losses = []
+        obs = torch.tensor(batch[SampleBatch.OBS], device=self.device)
+        actions = torch.tensor(batch[SampleBatch.ACTIONS], device=self.device)
+        ps = torch.zeros([batch.count], device=self.device)
+        returns = torch.zeros([batch.count], device=self.device)
+        discounts = torch.zeros([batch.count], device=self.device)
+
+        # Neccessary if policy uses recurrent/attention model
+        num_state_inputs = 0
+        for k in batch.keys():
+            if k.startswith("state_in_"):
+                num_state_inputs += 1
+        state_keys = ["state_in_{}".format(i) for i in range(num_state_inputs)]
+
+        # get rewards, old_prob, new_prob
+        rewards = batch[SampleBatch.REWARDS]
+        old_log_prob = torch.tensor(batch[SampleBatch.ACTION_LOGP])
+        new_log_prob = self.policy.compute_log_likelihoods(
+            actions=batch[SampleBatch.ACTIONS],
+            obs_batch=batch[SampleBatch.OBS],
+            state_batches=[batch[k] for k in state_keys],
+            prev_action_batch=batch.get(SampleBatch.PREV_ACTIONS),
+            prev_reward_batch=batch.get(SampleBatch.PREV_REWARDS),
+            actions_normalized=False,
+        ).detach()
+        prob_ratio = torch.exp(new_log_prob - old_log_prob)
+
+        eps_begin = 0
+        for episode in batch.split_by_episode():
+            eps_end = eps_begin + episode.count
+
+            # calculate importance ratios and returns
+
+            for t in range(episode.count):
+                discounts[eps_begin + t] = self.gamma ** t
+                if t == 0:
+                    pt_prev = 1.0
+                else:
+                    pt_prev = ps[eps_begin + t - 1]
+                ps[eps_begin + t] = pt_prev * prob_ratio[eps_begin + t]
+
+                # O(n^3)
+                # ret = 0
+                # for t_prime in range(t, episode.count):
+                #     gamma = self.gamma ** (t_prime - t)
+                #     rho_t_1_t_prime = 1.0
+                #     for k in range(t + 1, min(t_prime + 1, episode.count)):
```

Comment:
The earlier version of QReg was wrong here: k should start at t + 1, not t, and go all the way to t_prime (the subscripts in the [COBS](https://arxiv.org/pdf/1911.06854.pdf) paper are inclusive for the last index, not exclusive).